common/core/src/promise_utils.ts

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+'use strict';
+
+/**
+ * Defines type describing callback with two results.
+ *
+ * @template TResult1 - Type of the first result value, possibly Error.
+ * @template TResult2 - Type of the second result value, possible Error.
+ */
+export type DoubleValueCallback<TResult1, TResult2> = (result1?: TResult1, result2?: TResult2) => void;
+
+/**
+ * Defines type describing callback with one non-error result value.
+ *
+ * @template TResult - Type of the result value.
+ */
+export type NoErrorCallback<TResult> = (result?: TResult) => void;
+
+/**
+ * Defines type describing callback with only Error result value.
+ */
+export type ErrorCallback = (error?: Error) => void;
+
+/**
+ * Defines type describing regular callback with two results - one is the result value, the other one is Error.
+ *
+ * @template TResult - Type of the result value.
+ */
+export type Callback<TResult> = DoubleValueCallback<Error, TResult>;
+
+/**
+ * Defines type describing regular callback with three results - two are the result values, the other one is Error.
+ *
+ * @template TResult1 - Type of the first result value.
+ * @template TResult2 - Type of the second result value.
+ */
+export type TripleValueCallback<TResult1, TResult2> = (error?: Error, result1?: TResult1, result2?: TResult2) => void;
```

Comment:
I'm not a big fan of the name so if someone has better ideas I'd appreciate it. Same for `DoubleValueCallback`.